src/radix_connect/mobile/client.rs

```diff
@@ -1,65 +1,680 @@
+    #[uniffi::method]
+    pub async fn handle_deep_link(
+        &self,
+        url: String,
+    ) -> Result<RadixConnectMobileSessionRequest> {
+        // Try to parse the deep link ur to RadixConnectMobileRequest
+        let request = parse_mobile_connect_request(url)?;
+
+        // A session might be already established for the given session_id.
+        let existing_session = self.load_session(request.session_id).await.ok();
+        // Each new session requires origin validation.
+        let origin_requires_validation = existing_session.is_none();
+
+        match existing_session {
+            Some(session) => {
+                // Validate the request against the pre-validated session.
+                // This is the protection the Wallet can guarantee after the origin was validated by the user.
+                session.validate_request(&request)?;
+            }
+            None => {
+                // Don't yet store the session, as it requires validation from the user.
+                self.create_in_flight_session(&request)?;
+            }
+        }
+
+        // Return request to the Wallet to be handled by the user.
+        Ok(RadixConnectMobileSessionRequest::new(
+            request.session_id,
+            request.interaction,
+            request.origin,
+            origin_requires_validation,
+        ))
+    }
+
+    /// Send the Wallet's response to the dApp.
+    /// This is a stateful operation as it will save the session in the secure storage if the user validated the session.
+    #[uniffi::method]
+    pub async fn send_dapp_interaction_response(
+        &self,
+        wallet_response: RadixConnectMobileWalletResponse,
+    ) -> Result<()> {
+        let session_id = wallet_response.session_id;
+
+        // Get the in flight session, if any, that required validation from the user.
+        // Not remove at this stage, as the user might retry the interaction.
+        let in_flight_session = self
+            .new_sessions
+            .read()
+            .ok()
+            .and_then(|sessions| sessions.get(&session_id).cloned());
+
+        // Get the existing session, if any.
+        let existing_session =
+            self.load_session(wallet_response.session_id).await.ok();
+
+        let is_in_flight_session = in_flight_session.is_some();
+        let session = in_flight_session.or(existing_session).ok_or(
+            CommonError::RadixConnectMobileSessionNotFound { session_id },
+        )?;
+
+        let is_success_response = wallet_response.response.is_success();
+
+        // Send the wallet interaction response to the dApp through the transport.
+        self.wallet_interactions_transport
+            .send_wallet_interaction_response(
+                session.clone(),
+                wallet_response.response,
+            )
+            .await?;
+
+        _ = self.new_sessions.try_write().map(|mut new_sessions| {
```

Comment:
yes, failing to delete an in_fligh_session is not an error that should halt this function. If it happens to not be deleted, it will be ignored as the saved session takes priority.